pkg/odo/cli/init/init.go

```diff
@@ -0,0 +1,306 @@
+
+var initExample = templates.Examples(`
+  # Boostrap a new project in interactive mode
+  %[1]s
+`)
+
+type InitOptions struct {
+	// Backends to build init parameters
+	backends []params.ParamsBuilder
+
+	// filesystem on which command is running
+	fsys filesystem.Filesystem
+
+	// Clients
+	preferenceClient preference.Client
+
+	// the parameters needed to run the init procedure
+	params.InitParams
+
+	// Destination directory
+	destDir string
+}
+
+// NewInitOptions creates a new InitOptions instance
+func NewInitOptions(backends []params.ParamsBuilder, fsys filesystem.Filesystem, prefClient preference.Client) *InitOptions {
+	return &InitOptions{
+		backends:         backends,
+		fsys:             fsys,
+		preferenceClient: prefClient,
+		destDir:          ".",
+	}
+}
+
+// Complete will build the parameters for init, using different backends based on the flags set,
+// either by using flags or interactively is no flag is passed
+// Complete will return an error immediately if the current working directory is not empty
+func (o *InitOptions) Complete(cmdline cmdline.Cmdline, args []string) (err error) {
+
+	empty, err := isEmpty(o.fsys, o.destDir)
+	if err != nil {
+		return err
+	}
+	if !empty {
+		return errors.New("The current directory is not empty. You can bootstrap new component only in empty directory.\nIf you have existing code that you want to deploy use `odo deploy` or use `odo dev` command to quickly iterate on your component.")
+	}
+
+	flags := cmdline.GetFlags()
+	done := false
+	for _, backend := range o.backends {
+		if backend.IsAdequate(flags) {
+			o.InitParams, err = backend.ParamsBuild()
+			if err != nil {
+				return err
+			}
+			done = true
+			break
+		}
+	}
+
+	if !done {
+		odoutil.LogErrorAndExit(nil, "no backend found to build init parameters. This should not happen")
+	}
+	return nil
+}
+
+func isEmpty(fsys filesystem.Filesystem, path string) (bool, error) {
+	files, err := fsys.ReadDir(path)
+	if err != nil {
+		return false, err
+	}
+	return len(files) == 0, nil
+}
+
+// Validate validates the InitOptions based on completed values
+func (o *InitOptions) Validate() error {
+	return o.InitParams.Validate(o.preferenceClient)
+}
+
+// Run contains the logic for the odo command
+func (o *InitOptions) Run() (err error) {
+
+	var starterDownloaded bool
+
+	defer func() {
+		if err == nil {
+			return
+		}
+		if starterDownloaded {
+			err = fmt.Errorf("%w\nThe command failed after downloading the starter project. By security, the directory is not cleaned up.", err)
+		} else {
+			_ = o.fsys.Remove("devfile.yaml")
+			err = fmt.Errorf("%w\nThe command failed, the devfile has been removed from current directory.", err)
+		}
+	}()
+
+	destDevfile := filepath.Join(o.destDir, "devfile.yaml")
+	if o.InitParams.DevfilePath != "" {
+		err = o.downloadDirect(o.InitParams.DevfilePath, destDevfile)
+	} else {
+		err = o.downloadRegistry(o.InitParams.DevfileRegistry, o.InitParams.Devfile, o.destDir)
+	}
```

Comment:
Can the check `if o.InitParams.Starter != ""` be included in this block? I'm suggesting this because if the user asked for a starter project, the work done by any of these calls would be nullified. And they are http calls pulling things from the Internet. I think it would make the experience a bit faster. WDYT?

Reply:
If the user asked for a starter, we need to get the devfile to get the details about the starter (the URL, etc).

I agree that this part with the starter download deleting the devfile is quite complex, I cannot find a clean way to write it

Reply:
OK. Let's not fret over it for now, since we anyway need to download stuff.